helix-core/src/main/java/org/apache/helix/common/caches/ParticipantStateCache.java

```diff
@@ -0,0 +1,203 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import com.google.common.collect.Maps;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.helix.HelixDataAccessor;
+import org.apache.helix.PropertyKey;
+import org.apache.helix.common.controllers.ControlContextProvider;
+import org.apache.helix.controller.LogUtil;
+import org.apache.helix.model.LiveInstance;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public abstract class ParticipantStateCache<T> extends AbstractDataCache {
+  private static Logger LOG = LoggerFactory.getLogger(ParticipantStateCache.class.getName());
+  protected Map<String, Map<String, Map<String, T>>> _participantStateMap;
+
+  protected Map<PropertyKey, T> _participantStateCache = Maps.newHashMap();
+  public ParticipantStateCache(ControlContextProvider controlContextProvider) {
+    super(controlContextProvider);
+    _participantStateMap = Collections.emptyMap();
+  }
+
+  public ParticipantStateCache(String clusterName) {
+    this(createDefaultControlContextProvider(clusterName));
+  }
+
+
+  /**
+   * This refreshes the participant state cache data by re-fetching the data from zookeeper in an efficient way
+   *
+   * @param accessor
+   * @param liveInstanceMap map of all liveInstances in cluster
+   *
+   * @return
+   */
+  public boolean refresh(HelixDataAccessor accessor,
+      Map<String, LiveInstance> liveInstanceMap, Boolean snapshotEnabled, Set<String> restrictedKeys) {
+    long startTime = System.currentTimeMillis();
+
+    refreshParticipantStatesCacheFromZk(accessor, liveInstanceMap, snapshotEnabled, restrictedKeys);
+    Map<String, Map<String, Map<String, T>>> allParticipantStateMap = new HashMap<>();
+    for (PropertyKey key : _participantStateCache.keySet()) {
+      T participantState = _participantStateCache.get(key);
+      String[] params = key.getParams();
+      if (participantState != null && params.length >= 4) {
+        String keyLevel1 = params[1];
+        String keyLevel2 = params[2];
+        String keyLevel3 = params[3];
+        Map<String, Map<String, T>> stateMapLevel1 =
+            allParticipantStateMap.get(keyLevel1);
+        if (stateMapLevel1 == null) {
+          stateMapLevel1 = Maps.newHashMap();
+          allParticipantStateMap.put(keyLevel1, stateMapLevel1);
+        }
+        Map<String, T> stateMapLevel2 = stateMapLevel1.get(keyLevel2);
+        if (stateMapLevel2 == null) {
+          stateMapLevel2 = Maps.newHashMap();
+          stateMapLevel1.put(keyLevel2, stateMapLevel2);
+        }
+        stateMapLevel2.put(keyLevel3, participantState);
+      }
+    }
+
+    for (String instance : allParticipantStateMap.keySet()) {
```

Comment:
Since you are creating a new map every time already, is it still necessary to make it unmodifiable?